File: app/src/main/java/com/popgamma/tutor/QuestionBank.kt
package com.popgamma.tutor

// Pure Kotlin -- no Android imports.
//
// Finding 5 (plan): the spec's routing table crosses confidence with correctness but never says
// where correctness comes from. A hardcoded bank with deterministic string matching is the only
// way to force "confident wrong" on cue in front of a reviewer -- asking the model to judge
// correctness reintroduces exactly the variance a live demo can't afford.
//
// Finding 6 (plan): word-level timestamps can degrade transcription accuracy, and correctness is
// a substring match against the transcript -- so a single mistranscribed answer word flips
// `correct?` and silently selects a different routing-table row. The mitigation is here, in the
// bank content, not in code: every expected answer is short and phonetically distinct (a number
// or a single common word), never jargon or a near-homophone.
//
// Verified live against the API: Gemini's transcription normalizes spoken numbers to digits even
// in verbatim mode -- saying "twelve" comes back as "12." in the transcript. isCorrect() below
// accepts both forms for numeric answers so a correct answer never fails on formatting alone.
data class BankQuestion(val prompt: String, val expectedAnswer: String)

object QuestionBank {
    val questions = listOf(
        BankQuestion("What is seven plus five?", "twelve"),
        BankQuestion("What is nine times three?", "twenty seven"),
        BankQuestion("What is the capital of France?", "paris"),
        BankQuestion("What color do you get mixing blue and yellow?", "green"),
    )

    fun isCorrect(question: BankQuestion, transcript: String): Boolean {
        fun normalize(s: String) = s.lowercase()
            .replace('-', ' ')
            .replace(Regex("""\b(\d+)(st|nd|rd|th)\b"""), "$1")
            .replace("twelfth", "twelve")
            .replace(Regex("[^a-z0-9 ]"), "")
            .replace(Regex("\\s+"), " ")
            .trim()
        val normalizedTranscript = normalize(transcript)
        val normalizedExpected = normalize(question.expectedAnswer)
        if (Regex("\\b" + Regex.escape(normalizedExpected) + "\\b").containsMatchIn(normalizedTranscript)) return true

        // Verified against the live API: Gemini's transcription normalizes spoken numbers to
        // digits even in verbatim mode ("twelve" -> "12."), so a correct answer to a numeric
        // question would otherwise fail this match on formatting alone, not accuracy. Also accept
        // the digit form.
        val digitForm = wordsToDigits(question.expectedAnswer) ?: return false
        return Regex("\\b" + Regex.escape(digitForm) + "\\b").containsMatchIn(normalizedTranscript)
    }

    // 0-99 only, hardcoded -- the bank never needs more than this, not a general number parser.
    private val ONES = mapOf(
        "zero" to 0, "one" to 1, "two" to 2, "three" to 3, "four" to 4, "five" to 5,
        "six" to 6, "seven" to 7, "eight" to 8, "nine" to 9, "ten" to 10,
        "eleven" to 11, "twelve" to 12, "thirteen" to 13, "fourteen" to 14, "fifteen" to 15,
        "sixteen" to 16, "seventeen" to 17, "eighteen" to 18, "nineteen" to 19
    )
    private val TENS = mapOf(
        "twenty" to 20, "thirty" to 30, "forty" to 40, "fifty" to 50,
        "sixty" to 60, "seventy" to 70, "eighty" to 80, "ninety" to 90
    )

    private fun wordsToDigits(phrase: String): String? {
        val words = phrase.lowercase().trim().split(Regex("\\s+"))
        return when {
            words.size == 1 && ONES.containsKey(words[0]) -> ONES.getValue(words[0]).toString()
            words.size == 1 && TENS.containsKey(words[0]) -> TENS.getValue(words[0]).toString()
            words.size == 2 && TENS.containsKey(words[0]) && ONES[words[1]]?.let { it in 1..9 } == true ->
                (TENS.getValue(words[0]) + ONES.getValue(words[1])).toString()
            else -> null
        }
    }
}

File: web/src/logic/questionBank.ts
import type { BankQuestion } from '../types';

const ONES: Record<string, number> = {
  zero: 0, one: 1, two: 2, three: 3, four: 4, five: 5,
  six: 6, seven: 7, eight: 8, nine: 9, ten: 10,
  eleven: 11, twelve: 12, thirteen: 13, fourteen: 14, fifteen: 15,
  sixteen: 16, seventeen: 17, eighteen: 18, nineteen: 19
};

const TENS: Record<string, number> = {
  twenty: 20, thirty: 30, forty: 40, fifty: 50,
  sixty: 60, seventy: 70, eighty: 80, ninety: 90
};

function wordsToDigits(phrase: string): string | null {
  const words = phrase.toLowerCase().trim().split(/\s+/);
  if (words.length === 1 && ONES[words[0]] !== undefined) {
    return ONES[words[0]].toString();
  }
  if (words.length === 1 && TENS[words[0]] !== undefined) {
    return TENS[words[0]].toString();
  }
  if (words.length === 2 && TENS[words[0]] !== undefined && ONES[words[1]] && ONES[words[1]] >= 1 && ONES[words[1]] <= 9) {
    return (TENS[words[0]] + ONES[words[1]]).toString();
  }
  return null;
}

export const QuestionBank = {
  questions: [
    { prompt: 'What is seven plus five?', expectedAnswer: 'twelve' },
    { prompt: 'What is nine times three?', expectedAnswer: 'twenty seven' },
    { prompt: 'What is the capital of France?', expectedAnswer: 'paris' },
    { prompt: 'What color do you get mixing blue and yellow?', expectedAnswer: 'green' },
  ] as BankQuestion[],

  isCorrect(question: BankQuestion, transcript: string): boolean {
    const normalize = (s: string) =>
      s.toLowerCase()
        .replace(/-/g, ' ')
        .replace(/\b(\d+)(st|nd|rd|th)\b/g, '$1')
        .replace(/\btwelfth\b/g, 'twelve')
        .replace(/[^a-z0-9 ]/g, '')
        .replace(/\s+/g, ' ')
        .trim();

    const normalizedTranscript = normalize(transcript);
    const normalizedExpected = normalize(question.expectedAnswer);

    const expectedRegex = new RegExp(`\\b${normalizedExpected}\\b`, 'i');
    if (expectedRegex.test(normalizedTranscript)) {
      return true;
    }

    // Accept digit form (e.g., "12." for "twelve")
    const digitForm = wordsToDigits(question.expectedAnswer);
    if (digitForm) {
      const digitRegex = new RegExp(`\\b${digitForm}\\b`);
      if (digitRegex.test(normalizedTranscript)) {
        return true;
      }
    }

    return false;
  }
};
